Refactor Feature and Parser

gchoqueux opened the pull request:
## Description
* replace parsing options `withNormal` and `withAltitude` by `structure`.
* simplify build extent check in `Feature`.
* normalize crs projection `FeatureCollection`.
* remove `optionsFeature` from `FeatureCollection`.
* simplify  `FeatureCollection` and `Feature` constructor.

## BREAKING CHANGE

`FeatureCollection` and `Feature` signature constructor are changed.

## DEPRECATED

Parsing options `withNormal` and `withAltitude` are deprecated, use `structure` (`2d` or `3d`) instead of.

gchoqueux committed:
refactor(Feature): normalize crs projection.

gchoqueux committed:
refactor(Feature): simplify build extent check.

gchoqueux committed:
test(unit): update unit tests with feature refactoring.

mgermerie commented on src/Core/Feature.js:
```diff
@@ -195,29 +195,23 @@ class Feature {
-     * @param {boolean} [options.withNormal] Set vertice normal when adding new vertice.
-     * @param {Style} [options.style] The style to inherit when creating a new
-     * style for this feature.
+     * @param {FeatureCollection} [options={}] Parent feature collection.
```
Maybe we could rename `options` to something like `parent` or `collection` ?

gchoqueux replied:
yes, i rename to `collection`

mgermerie commented on src/Core/Feature.js:
```diff
@@ -27,10 +28,9 @@ const defaultNormal = new THREE.Vector3(0, 0, 1);
- * True if the layer inherits from {@link GeometryLayer}
- * @property {boolean} [withAltitude=true] - If true each coordinate altitude is kept
- * True if the layer inherits from {@link GeometryLayer}
+ * @property {string} [structure='3d'] - data structure type : 2d or 3d.
```
I believe the default `3d` value for `structure` parameter is not implemented in `FeatureCollection` constructor. If I am not mistaking, when a user does not pass a value for the `structure` parameter, the `FeatureCollection` will act as if it was `2d` (at line 277 : `this.size = options.structure == '3d' ? 3 : 2;`).

gchoqueux replied:
yes!, i change the doc to `2d`

mgermerie commented on examples/source_file_gpx_3d.html:
```diff
@@ -74,6 +74,7 @@
                             crs: view.referenceCrs,
                             buildExtent: true,
                             mergeFeatures: true,
+                            structure: '3d',
```
According to the documentation, `3d` is supposed to be the default value for `structure` parameter. As I mention in another comment, it is not yet implemented, but this line could be removed if it finally gets implemented.

gchoqueux replied:
doc changed

mgermerie commented on test/unit/feature2mesh.js:
```diff
@@ -27,8 +27,8 @@ function computeAreaOfMesh(mesh) {
 describe('Feature2Mesh', function () {
-    const parsed = GeoJsonParser.parse(geojson, { in: { crs: 'EPSG:3946' }, out: { crs: 'EPSG:3946', buildExtent: true, mergeFeatures: false } });
-    const parsed2 = GeoJsonParser.parse(geojson2, { in: { crs: 'EPSG:3946' }, out: { crs: 'EPSG:3946', buildExtent: true, mergeFeatures: false } });
+    const parsed = GeoJsonParser.parse(geojson, { in: { crs: 'EPSG:3946' }, out: { crs: 'EPSG:3946', buildExtent: true, mergeFeatures: false, structure: '3d' } });
```
Same as previously mentionned : the `structure` parameter could be removed shall its default value be implemented.

gchoqueux replied:
doc changed

gchoqueux committed:
refactor(Feature): remove optionsFeature from FeatureCollection.

BREAKING CHANGE:

FeatureCollection and Feature signature constructor are changed.

gchoqueux committed:
refactor(Feature): replace parsing option withNormal and withAltitude by structure.

DEPRECATED:
withNormal and withAltitude are deprecated.
